### PR TITLE
Rename `SemIR::Node` to `SemIR::Inst`

### PR DESCRIPTION
**Pull request by josh11b**
And generally replace "node" by "inst" in the code and "instruction" in comments.

### CODE REVIEW

File: toolchain/check/handle_index.cpp [hunk 1]
```diff
@@ -20,60 +20,60 @@ auto HandleIndexExpressionStart(Context& /*context*/,
...
-  auto index_node = context.nodes().Get(index_node_id);
-  auto operand_node_id = context.node_stack().PopExpression();
-  operand_node_id =
-      ConvertToValueOrReferenceExpression(context, operand_node_id);
-  auto operand_node = context.nodes().Get(operand_node_id);
-  auto operand_type_id = operand_node.type_id();
-  auto operand_type_node = context.nodes().Get(
+  SemIR::InstId index_inst_id = context.node_stack().PopExpression();
+  auto index_inst = context.insts().Get(index_inst_id);
+  SemIR::InstId operand_inst_id = context.node_stack().PopExpression();
```
Comment (chandlerc) [inline toolchain/check/handle_index.cpp]: As discussed a bit on Discord, let's leave the `auto` usage consistent before and after here, and if we want to adjust, do that in a separate PR.

```suggestion
  auto index_inst_id = context.node_stack().PopExpression();
  auto index_inst = context.insts().Get(index_inst_id);
  auto operand_inst_id = context.node_stack().PopExpression();
```

File: toolchain/check/inst_block_stack.h
Comment (chandlerc) [inline toolchain/check/inst_block_stack.h:1]: This doesn't seem to be marked as a move in Git -- is it possible to get that to be marked so that there is history?

File: toolchain/parse/context.h [hunk 2]
```diff
@@ -114,7 +114,7 @@ class Context {
       -> void;
 
   // Adds a node to the parse tree that has children.
-  auto AddNode(NodeKind kind, Lex::Token token, int subtree_start,
+  auto AddInst(NodeKind kind, Lex::Token token, int subtree_start,
```
Comment (chandlerc) [inline toolchain/parse/context.h]: I think this change (and all the others that make it compile) is wrong?

File: toolchain/sem_ir/formatter.cpp [hunk 3]
```diff
@@ -296,14 +297,14 @@ class NodeNamer {
                                   *this, parse_node, std::move(name))};
   }
 
-  // Finds and adds a suitable block label for the given semantics node that
+  // Finds and adds a suitable block label for the given sem_ir instruction that
```
Comment (chandlerc) [inline toolchain/sem_ir/formatter.cpp]: Slight preference for:
```suggestion
  // Finds and adds a suitable block label for the given SemIR instruction that
```

Or:
```suggestion
  // Finds and adds a suitable block label for the given semantics IR instruction that
```

File: toolchain/sem_ir/inst_kind.cpp
Comment (chandlerc) [inline toolchain/sem_ir/inst_kind.cpp:1]: Missing history across rename.

File: toolchain/sem_ir/inst_kind.def
Comment (chandlerc) [inline toolchain/sem_ir/inst_kind.def:1]: Missing history across rename.

File: toolchain/sem_ir/inst_kind.h
Comment (chandlerc) [inline toolchain/sem_ir/inst_kind.h:1]: Missing history across rename.

File: toolchain/sem_ir/typed_insts_test.cpp
Comment (chandlerc) [inline toolchain/sem_ir/typed_insts_test.cpp:1]: Missing history across rename.

File: toolchain/check/testdata/basics/builtin_insts.carbon
Comment (chandlerc) [inline toolchain/check/testdata/basics/builtin_insts.carbon:1]: Missing history across rename.

File: toolchain/check/inst_block_stack.cpp
Comment (chandlerc) [inline toolchain/check/inst_block_stack.cpp:1]: Just noting that the history worked well here (and few other places). I've tried to leave file comments for all the files that seem to have lost history across rename.

File: toolchain/check/handle_function.cpp [hunk 4]
```diff
@@ -49,29 +49,29 @@ static auto BuildFunctionDeclaration(Context& context, bool is_definition)
...
       context.node_stack().Pop<Parse::NodeKind::ParameterList>();
-  SemIR::NodeBlockId implicit_param_refs_id =
+  SemIR::InstBlockId implicit_param_refs_id =
       context.node_stack()
           .PopIf<Parse::NodeKind::ImplicitParameterList>()
-          .value_or(SemIR::NodeBlockId::Empty);
+          .value_or(SemIR::InstBlockId::Empty);
   auto name_context = context.declaration_name_stack().Pop();
-  auto fn_node =
+  Parse::Node fn_node =
```
Comment (chandlerc) [inline toolchain/check/handle_function.cpp]: Let's preserve `auto` for now:
```suggestion
  auto fn_node =
```

File: toolchain/check/handle_call_expression.cpp [hunk 5]
```diff
@@ -78,10 +78,10 @@ auto HandleCallExpression(Context& context, Parse::Node parse_node) -> bool {
                       context.params_or_args_stack().PeekCurrentBlockContents(),
                       return_storage_id, function_decl->parse_node,
                       callable.implicit_param_refs_id, callable.param_refs_id);
-  auto call_node_id = context.AddNode(
+  SemIR::InstId call_inst_id = context.AddInst(
```
Comment (chandlerc) [inline toolchain/check/handle_call_expression.cpp]: Preserving `auto` for now:
```suggestion
  auto call_inst_id = context.AddInst(
```

File: toolchain/check/handle_array.cpp [hunk 6]
```diff
@@ -30,26 +30,26 @@ auto HandleArrayExpression(Context& context, Parse::Node parse_node) -> bool {
...
   }
 
-  auto bound_node_id = context.node_stack().PopExpression();
+  SemIR::InstId bound_inst_id = context.node_stack().PopExpression();
   context.node_stack()
       .PopAndDiscardSoloParseNode<Parse::NodeKind::ArrayExpressionSemi>();
-  auto element_type_node_id = context.node_stack().PopExpression();
-  auto bound_node = context.nodes().Get(bound_node_id);
-  if (auto literal = bound_node.TryAs<SemIR::IntegerLiteral>()) {
+  SemIR::InstId element_type_inst_id = context.node_stack().PopExpression();
```
Comment (chandlerc) [inline toolchain/check/handle_array.cpp]: Preserve `auto` for now:
```suggestion
  auto element_type_inst_id = context.node_stack().PopExpression();
```

File: toolchain/check/handle_array.cpp [hunk 7]
```diff
@@ -30,26 +30,26 @@ auto HandleArrayExpression(Context& context, Parse::Node parse_node) -> bool {
     return context.TODO(parse_node, "HandleArrayExpressionWithoutBounds");
   }
 
-  auto bound_node_id = context.node_stack().PopExpression();
+  SemIR::InstId bound_inst_id = context.node_stack().PopExpression();
```
Comment (chandlerc) [inline toolchain/check/handle_array.cpp]: Preserve `auto` for now:
```suggestion
  auto bound_inst_id = context.node_stack().PopExpression();
```

File: toolchain/check/handle_call_expression.cpp (hunk 5 above)
Comment (chandlerc) [inline toolchain/check/handle_call_expression.cpp]: Preserving `auto` for now:
```suggestion
  auto call_inst_id = context.AddInst(
```

File: toolchain/check/declaration_name_stack.cpp [hunk 8]
```diff
@@ -73,15 +73,15 @@ auto DeclarationNameStack::LookupOrAddName(NameContext name_context,
...
   }
 }
 
 auto DeclarationNameStack::AddNameToLookup(NameContext name_context,
-                                           SemIR::NodeId target_id) -> void {
-  auto existing_node_id = LookupOrAddName(name_context, target_id);
-  if (existing_node_id.is_valid()) {
-    context_->DiagnoseDuplicateName(name_context.parse_node, existing_node_id);
+                                           SemIR::InstId target_id) -> void {
+  SemIR::InstId existing_inst_id = LookupOrAddName(name_context, target_id);
```
Comment (chandlerc) [inline toolchain/check/declaration_name_stack.cpp]: Preserve `auto` for now:
```suggestion
  auto existing_inst_id = LookupOrAddName(name_context, target_id);
```

File: toolchain/check/declaration_name_stack.cpp [hunk 9]
```diff
@@ -95,22 +95,23 @@ auto DeclarationNameStack::ApplyNameQualifierTo(NameContext& name_context,
...
   if (CanResolveQualifier(name_context, parse_node)) {
     // For identifier nodes, we need to perform a lookup on the identifier.
-    // This means the input node_id is actually a string ID.
+    // This means the input instruction name_id is actually a string ID.
     //
     // TODO: This doesn't perform the right kind of lookup. We will find names
     // from enclosing lexical scopes here, in the case where `target_scope_id`
     // is invalid.
-    auto resolved_node_id = context_->LookupName(
+    SemIR::InstId resolved_inst_id = context_->LookupName(
```
Comment (chandlerc) [inline toolchain/check/declaration_name_stack.cpp]: Preserve `auto` for now:
```suggestion
    auto resolved_inst_id = context_->LookupName(
```

Comment (chandlerc) [inline toolchain/check/declaration_name_stack.cpp]: Preserve `auto` for now:
```suggestion
    auto resolved_inst_id = context_->LookupName(
```

File: toolchain/check/handle_call_expression.cpp (hunk 5 above)
Comment (chandlerc) [inline toolchain/check/handle_call_expression.cpp]: Preserving `auto` for now:
```suggestion
  auto call_inst_id = context.AddInst(
```

File: toolchain/check/context.h [hunk 10]
```diff
@@ -84,64 +84,65 @@ class Context {
 
   // Returns the current scope, if it is of the specified kind. Otherwise,
   // returns nullopt.
-  template <typename NodeT>
-  auto GetCurrentScopeAs() -> std::optional<NodeT> {
-    auto current_scope_node_id = current_scope().scope_node_id;
-    if (!current_scope_node_id.is_valid()) {
+  template <typename InstT>
+  auto GetCurrentScopeAs() -> std::optional<InstT> {
+    SemIR::InstId current_scope_inst_id = current_scope().scope_inst_id;
```
Comment (chandlerc) [inline toolchain/check/context.h]: Preserve `auto` for now:
```suggestion
    auto current_scope_inst_id = current_scope().scope_inst_id;
```

File: toolchain/check/context.cpp [hunk 11]
```diff
@@ -373,14 +373,14 @@ class TypeCompleter {
       return true;
     }
 
-    auto node_id = context_.sem_ir().GetTypeAllowBuiltinTypes(type_id);
-    auto node = context_.nodes().Get(node_id);
+    SemIR::InstId inst_id = context_.sem_ir().GetTypeAllowBuiltinTypes(type_id);
```
Comment (chandlerc) [inline toolchain/check/context.cpp]: Preserve `auto` for now:
```suggestion
    auto inst_id = context_.sem_ir().GetTypeAllowBuiltinTypes(type_id);
```

File: toolchain/check/handle_name.cpp [hunk 12]
```diff
@@ -44,38 +44,39 @@ auto HandleMemberAccessExpression(Context& context, Parse::Node parse_node)
...
   // into that scope.
   if (auto name_scope_id = GetAsNameScope(context, base_id)) {
-    auto node_id = name_scope_id->is_valid()
-                       ? context.LookupName(parse_node, name_id, *name_scope_id,
-                                            /*print_diagnostics=*/true)
-                       : SemIR::NodeId::BuiltinError;
-    auto node = context.nodes().Get(node_id);
-    // TODO: Track that this node was named within `base_id`.
-    context.AddNodeAndPush(
+    SemIR::InstId inst_id =
```
Comment (chandlerc) [inline toolchain/check/handle_name.cpp]: Preserve `auto` for now:
```suggestion
    auto inst_id =
```

File: toolchain/check/handle_pattern_binding.cpp [hunk 13]
```diff
@@ -77,29 +77,31 @@ auto HandlePatternBinding(Context& context, Parse::Node parse_node) -> bool {
...
       SemIR::TypeId value_type_id = cast_type_id;
       if (enclosing_class_decl) {
         auto& class_info =
             context.classes().Get(enclosing_class_decl->class_id);
-        auto field_type_node_id = context.AddNode(SemIR::UnboundFieldType{
-            parse_node, context.GetBuiltinType(SemIR::BuiltinKind::TypeType),
-            class_info.self_type_id, cast_type_id});
-        value_type_id = context.CanonicalizeType(field_type_node_id);
-        value_id = context.AddNode(
+        SemIR::InstId field_type_inst_id =
```
Comment (chandlerc) [inline toolchain/check/handle_pattern_binding.cpp]: Preserve `auto` for now:
```suggestion
        auto field_type_inst_id =
```

Comment (chandlerc) [inline toolchain/check/handle_pattern_binding.cpp]: Preserve `auto` for now:
```suggestion
        auto field_type_inst_id =
```

Comment (chandlerc) [inline toolchain/check/handle_pattern_binding.cpp]: Preserve `auto` for now:
```suggestion
        auto field_type_inst_id =
```

File: toolchain/check/pending_block.h [hunk 14]
```diff
@@ -39,52 +39,52 @@ class PendingBlock {
     size_t size_;
   };
 
-  auto AddNode(SemIR::Node node) -> SemIR::NodeId {
-    auto node_id = context_.nodes().AddInNoBlock(node);
-    nodes_.push_back(node_id);
-    return node_id;
+  auto AddInst(SemIR::Inst inst) -> SemIR::InstId {
+    SemIR::InstId inst_id = context_.insts().AddInNoBlock(inst);
```
Comment (chandlerc) [inline toolchain/check/pending_block.h]: Preserve `auto` for now:
```suggestion
    auto inst_id = context_.insts().AddInNoBlock(inst);
```

File: toolchain/sem_ir/typed_insts_test.cpp
Comment (chandlerc) [inline toolchain/sem_ir/typed_insts_test.cpp:1]: Missing history across rename.

**Review by chandlerc — COMMENTED** [review]
See comments inline, I think mostly with suggestions you can just apply.

I think one big change will just be to revert the changes to `toolchain/parse` which seems likely a scripting mis-fire.

I haven't yet reviewed the files that didn't have history as that's really important to make the review meaningful here.

If the individual changes are too hard to apply individually, or you get stuck with a bunch of merge conflicts, it may make sense to start with a more minimal fully automated change and separate the manual tweaks into follow-up PRs.

File: toolchain/parse/context.h (hunk 2 above)
Comment (josh11b) [reply toolchain/parse/context.h]: Undone.

Comment (josh11b) [reply toolchain/parse/context.h]: Undone.

**Review by chandlerc — APPROVED** [review]
looks good, thanks for fixing up all the history!
